### PR TITLE
Fix PR part 3 for multi-gpu fix infrastructure, adding context object for collecting relevant tensors and centralized forward pass logic

### PR DESCRIPTION
**Pull request by wli51**
## Overview
This PR is part 3 for fixing multi-GPU support. It extends the design implemented in #17, which centralizes loss computation as its own module, to also isolate forward passing logic away from the trainer implementation. The outcome of this PR is that:

instead of needing this in the trainer,
which gets even more complicated with GAN training where there are two models
```python
class Trainer:
    ...
    def train():
        for batch in self.dataloader:
            y_pred = self.model(batch)
            total_loss = 0
            for loss in losses:
                total_loss += loss(y_pred, batch)
            total_loss.backward()
            optimizer.step()
            
            # if there is a second model for gan ...
            self.model.eval()
            p = self.critic(self.model(batch))
            total_critic_loss = 0
            for loss in critic_losses:
                 total_critic_loss+= loss(p, batch)
            total_critic_loss.backward()
            critic_optimizer.step()
            ...
```

now the trainer would just look like
```python
class Trainer:
    ...
    def train():
        for batch in self.dataloader:
            for group in self.forward_groups: # each model has its defined group
                ctx = group(batch, train=True)
                total_loss, log = self.loss_group(ctx)
                total_loss.backward()
                optimizer.step()
            ...
```
regardless of how many models there are and what kind of topology the different models depend on each other

## Adds:
1. `context.py`: context object to organize different model outputs into one and perform some validation
2. `forward_groups.py`: module to house all forward operation definitions, currently has the simple one model realization
3. `names.py`: helper to define common names for outputs that gets produced that mean the samething across different model architectures (input image is always the same input image whether in a unet or in a gan)
4. `tests

## Refactors:
- Move the loss group implementation from merged #17 also into the engine folder for better organization
- with minor import fixes

### CODE REVIEW

**Review by axiomcura — APPROVED** [review]
LGTM! Nice PR c:

File: src/virtual_stain_flow/engine/forward_groups.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    """
+
+    # Subclasses should override these with ordered tuples.
+    input_keys:  Tuple[str, ...]
+    target_keys: Tuple[str, ...]
+    output_keys: Tuple[str, ...]
+
+    def __init__(
+        self,
+        device: torch.device = torch.device("cpu")
```
[the remark this answers is not in the document]
Comment (wli51) [reply src/virtual_stain_flow/engine/forward_groups.py]: gpu is all you need for compute indeed

[the remark this answers is not in the document]
Comment (wli51) [reply src/virtual_stain_flow/engine/forward_groups.py]: i don't think we are at risk here since the device param is used read only in the class

File: src/virtual_stain_flow/engine/forward_groups.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        # 1) Stage and validate inputs/targets
+        ctx = Context(**self._move_tensors(inputs), **{GENERATOR_MODEL: fp_model })
+        ctx.require(self.input_keys)
+        ctx.require(self.target_keys)
+
+        # 2) Train/Eval switches
+        if train:
+            fp_model.train()
+            if fp_optimizer is not None:
+                fp_optimizer.zero_grad(set_to_none=True)
```
[the remark this answers is not in the document]
Comment (wli51) [reply src/virtual_stain_flow/engine/forward_groups.py]: Yes I am seeing the lack of clarity of what `forward_groups`, `loss_group` and `trainer` modules each are responsible for in model training. Added richer doc strings in these modules to enhance clarity. 

In a nutshell:
1. `forward_groups`: takes care of optimizer zeroing, model state (and grad state) switching, returns an organized context object of forward pass products. In this process if grad state is train then gradients are actually tracked (but cannot be computed because loss has not been computed yet).
2. `loss_group`: takes care only of multiple loss item management, feeding of proper items from the context object to each loss function that can require different inputs, and producing a final weighted total loss that is back propagate-able if gradients are tracked.
3. `trainer`: orchestrator, calls forward_group and then loss_group to get the total loss, and then calls `total_loss.backward()` and  `optimizer.step()` in the end to back-prop and update the model weights.

**Comment by wli51** [comment]
Thanks @axiomcura. Merging, and cleaning residue coffee in your mug.
